# Assignment 1
### CSCN 8020_Reinforcement Learning Programming
### Vishnu Sivaraj
### 9025320

## Problem 1

We design the pick-and-place robot control task as a Markov Decision Process (MDP).  

An MDP is defined as:

(S, A, P, R, γ)

Where:
- S = State space
- A = Action space
- P = Transition function
- R = Reward function
- γ = Discount factor

The goal is to learn fast and smooth movements while successfully completing the pick-and-place task.

---

### 1. State Space (S)

The state must include all information required to make optimal decisions (Markov Property).

We define the state as:

s = (q₁, q₂, ..., qₙ, q̇₁, q̇₂, ..., q̇ₙ, x_obj, g)

Where:
- qᵢ = Joint angles (positions)
- q̇ᵢ = Joint velocities
- x_obj = Object position
- g = Grasp status (0 = not grasped, 1 = grasped)

### Reasoning:
- Joint positions and velocities are required for smooth motion control.
- Object position is needed for reaching and placing.
- Grasp status determines task progress.
- Including velocities ensures the system satisfies the Markov property.

This is a continuous state space.

---

### 2. Action Space (A)

Since the agent directly controls motors, actions are motor torques:

a = (τ₁, τ₂, ..., τₙ)

Where:
- τᵢ = Torque applied to joint i

### Reasoning:
- Direct torque control allows learning smooth and precise movements.
- This creates a continuous action space.

---

### 3. Transition Function (P)

The transition function describes how the robot moves:

sₜ₊₁ = f(sₜ, aₜ)

Where f represents robotic dynamics governed by:
- Physics
- Gravity
- Friction
- Motor dynamics

Transitions may be:
- Deterministic (in simulation)
- Stochastic (in real-world due to noise)

---

### 4. Reward Function (R)

To encourage fast, smooth, and successful behavior, the reward function includes:

### 1. Success Reward
+100 when the object is successfully placed at the target location.

### 2. Step Penalty
−1 per time step to encourage faster completion.

### 3. Smoothness Penalty
−λ ||τₜ − τₜ₋₁||²  
Penalizes large torque changes to encourage smooth motion.

### 4. Distance Penalty
−β × (distance between gripper and object)  
Encourages efficient reaching behavior.

### Final Reward Structure:

R = 
- +100 if successful placement  
- −1 per step  
- −λ ||τₜ − τₜ₋₁||²  
- −β × distance  

---

### 5. Discount Factor (γ)

We choose:

γ = 0.95 to 0.99

### Reasoning:
- Encourages long-term planning.
- Ensures the agent values completing the task.
- Maintains stability in learning.

---

### 6. Final MDP Summary

State (S):
Joint angles, joint velocities, object position, grasp status.

Action (A):
Continuous motor torques.

Transition (P):
Governed by robot dynamics and physics.

Reward (R):
Combination of success reward, time penalty, smoothness penalty, and distance penalty.

Discount (γ):
Close to 1 for long-term planning.

---

### 7. Conclusion

The pick-and-place robot problem is modeled as a continuous-state, continuous-action MDP.
Such problems are typically solved using modern deep reinforcement learning algorithms like:
- PPO
- SAC
- DDPG
- TD3

This formulation enables the robot to learn efficient, fast, and smooth motion control behavior.


### Reflection

In this problem, i modeled the pick-and-place robot task as a Markov Decision Process, which helps us structure how the robot learns from interaction. An MDP has five main components: states, actions, transitions, rewards, and a discount factor. 

The state represents everything the robot needs to make a good decision, including joint positions, joint velocities, the object’s position, and whether the object has been grasped.
We include velocities because smooth movement depends not just on where the joints are, but how fast they are moving. 

The actions are the motor torques applied to each joint, since the robot directly controls its motors for precise and smooth motion. The transition function represents how the robot moves according to physical dynamics. The reward function is the most important part because it teaches the robot what good behavior looks like. We give a large positive reward for successfully placing the object, a small penalty per step to encourage speed, and additional penalties to ensure smooth and efficient movement. The discount factor is chosen close to one so the robot values long-term success and completes the full task optimally.


## Problem 2

### 2×2 Gridworld – Two Iterations of Value Iteration

We are given a 2×2 gridworld with four states:

- States:  
  $$
  S = \{s_1, s_2, s_3, s_4\}
  $$  
  arranged as:
  - top-left: $s_1$
  - top-right: $s_2$
  - bottom-left: $s_3$
  - bottom-right: $s_4$

- Actions:  
  $$
  A = \{\text{up}, \text{down}, \text{left}, \text{right}\}
  $$

- Transition dynamics:
  - If an action would move the agent into a wall, the agent stays in the same state.
  - Otherwise, the agent moves deterministically to the neighboring state.

- Rewards (state-based):
  $$
  R(s_1)=5, \quad
  R(s_2)=10, \quad
  R(s_3)=1, \quad
  R(s_4)=2
  $$

- Discount factor:
  $$
  \gamma = 0.9
  $$

Bellman Optimality Update:

$$
V_{k+1}(s) = \max_{a \in A} \left[ R(s) + \gamma V_k(s'(s,a)) \right]
$$

---

## Iteration 1

Initial values:

$$
V_0(s) = 0 \quad \forall s
$$

Since all next-state values are zero:

$$
V_1(s) = R(s)
$$

Therefore:

- $V_1(s_1)=5$
- $V_1(s_2)=10$
- $V_1(s_3)=1$
- $V_1(s_4)=2$

---

## Iteration 2

We compute:

$$
V_2(s) = \max_{a \in A} \left[ R(s) + \gamma V_1(s'(s,a)) \right]
$$

---

### State $s_1$

Possible next states: $s_1, s_2, s_3$

\[
\begin{aligned}
Q(s_1,\text{up}) &= 5 + 0.9(5) = 9.5 \\
Q(s_1,\text{left}) &= 9.5 \\
Q(s_1,\text{right}) &= 5 + 0.9(10) = 14 \\
Q(s_1,\text{down}) &= 5 + 0.9(1) = 5.9
\end{aligned}
\]

$$
V_2(s_1)=14
$$

---

### State $s_2$

\[
\begin{aligned}
Q(s_2,\text{up}) &= 10 + 0.9(10) = 19 \\
Q(s_2,\text{right}) &= 19 \\
Q(s_2,\text{left}) &= 14.5 \\
Q(s_2,\text{down}) &= 11.8
\end{aligned}
\]

$$
V_2(s_2)=19
$$

---

### State $s_3$

\[
\begin{aligned}
Q(s_3,\text{down}) &= 1.9 \\
Q(s_3,\text{left}) &= 1.9 \\
Q(s_3,\text{up}) &= 5.5 \\
Q(s_3,\text{right}) &= 2.8
\end{aligned}
\]

$$
V_2(s_3)=5.5
$$

---

### State $s_4$

\[
\begin{aligned}
Q(s_4,\text{down}) &= 3.8 \\
Q(s_4,\text{right}) &= 3.8 \\
Q(s_4,\text{up}) &= 11 \\
Q(s_4,\text{left}) &= 2.9
\end{aligned}
\]

$$
V_2(s_4)=11
$$

---

## Final Values After Two Iterations

| State | $V_0(s)$ | $V_1(s)$ | $V_2(s)$ |
|-------|----------|----------|----------|
| s1    | 0        | 5        | 14       |
| s2    | 0        | 10       | 19       |
| s3    | 0        | 1        | 5.5      |
| s4    | 0        | 2        | 11       |

Thus,

- $V_2(s_1)=14$
- $V_2(s_2)=19$
- $V_2(s_3)=5.5$
- $V_2(s_4)=11$



## Reflection

In this problem, I learned how value iteration works step by step. At the beginning, all state values are zero, so during the first iteration the value of each state simply becomes its immediate reward. However, in the second iteration, we start to see how future rewards influence the value of each state.

The most interesting part is how the highest reward state, s2, starts affecting the neighboring states. Even though s1 and s4 do not have the highest rewards themselves, their values increase significantly because they can move toward s2. This shows how value iteration spreads reward information backward through the grid.

I also understood that the algorithm does not follow the initial policy given in the question. Instead, it always chooses the action that maximizes the expected return. Over multiple iterations, the values would continue to update until they stabilize, which would give the optimal value function.


## Problem 3

In this problem, we solve a 5×5 gridworld using Value Iteration. The goal is to find the optimal value function 
𝑉∗
 and optimal policy 
π∗
 based on the given reward rules. The agent receives +10 when it reaches the goal state, -5 in grey penalty states, and -1 in all other normal states. We implemented both standard value iteration and in-place value iteration and compared their performance.

In [2]:
%pip install numpy

  Using cached numpy-2.4.2-cp312-cp312-win_amd64.whl.metadata (6.6 kB)
Using cached numpy-2.4.2-cp312-cp312-win_amd64.whl (12.3 MB)
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
# Task 1: Reward function + Standard Value Iteration
# Task 2: In-Place Value Iteration + performance comparison

import numpy as np
import time

# -----------------------------
# 1) Environment specification
# -----------------------------
ROWS, COLS = 5, 5
ACTIONS = ["R", "D", "L", "U"]
ACTION_TO_DELTA = {"R": (0, 1), "D": (1, 0), "L": (0, -1), "U": (-1, 0)}

GAMMA = 0.9
THETA = 0.05  

GOAL = (4, 4)
GREY_STATES = {(2, 2), (3, 0), (0, 4)}

def in_bounds(r, c):
    return 0 <= r < ROWS and 0 <= c < COLS

def step(state, action):
    """Deterministic move; if hit wall, stay."""
    r, c = state
    dr, dc = ACTION_TO_DELTA[action]
    nr, nc = r + dr, c + dc
    if in_bounds(nr, nc):
        return (nr, nc)
    return (r, c)

# Reward function exactly as the question states: R(s)
REWARDS = np.full((ROWS, COLS), -1.0)
for (r, c) in GREY_STATES:
    REWARDS[r, c] = -5.0
REWARDS[GOAL] = 10.0

# IMPORTANT (matches typical RL convention + question):
# We treat reward as the reward of the NEXT state: r = R(s')
# and terminal state's value is fixed to 0 because episode ends there.
# This uses the given R(s) exactly, but avoids "double counting" at the goal.
def bellman_backup(V, s, gamma=GAMMA):
    """Compute max_a [ R(s') + gamma * V(s') ]"""
    r, c = s
    q_vals = []
    for a in ACTIONS:
        ns = step((r, c), a)
        nr, nc = ns
        q_vals.append(REWARDS[nr, nc] + gamma * V[nr, nc])
    return max(q_vals)

def extract_greedy_policy(V, gamma=GAMMA, tol=1e-9):
    """Return policy grid with ties shown as e.g. 'R|D'. Terminal shown as 'G'."""
    policy = np.empty((ROWS, COLS), dtype=object)
    for r in range(ROWS):
        for c in range(COLS):
            if (r, c) == GOAL:
                policy[r, c] = "G"
                continue
            q = {}
            best = -1e18
            for a in ACTIONS:
                nr, nc = step((r, c), a)
                q[a] = REWARDS[nr, nc] + gamma * V[nr, nc]
                best = max(best, q[a])
            best_actions = [a for a in ACTIONS if abs(q[a] - best) <= tol]
            policy[r, c] = "|".join(best_actions)
    return policy

def value_iteration_copy(gamma=GAMMA, theta=THETA):
    """Standard value iteration using two arrays (copy-based)."""
    V = np.zeros((ROWS, COLS), dtype=float)
    V[GOAL] = 0.0  # terminal fixed

    sweeps = 0
    t0 = time.perf_counter()

    while True:
        sweeps += 1
        V_new = V.copy()
        delta = 0.0

        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) == GOAL:
                    V_new[r, c] = 0.0
                    continue
                old = V[r, c]
                V_new[r, c] = bellman_backup(V, (r, c), gamma=gamma)
                delta = max(delta, abs(V_new[r, c] - old))

        V = V_new
        if delta < theta:
            break

    elapsed = time.perf_counter() - t0
    pi = extract_greedy_policy(V, gamma=gamma)
    return V, pi, sweeps, elapsed

def value_iteration_inplace(gamma=GAMMA, theta=THETA):
    """In-place value iteration using one array (immediate updates)."""
    V = np.zeros((ROWS, COLS), dtype=float)
    V[GOAL] = 0.0  # terminal fixed

    sweeps = 0
    t0 = time.perf_counter()

    while True:
        sweeps += 1
        delta = 0.0

        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) == GOAL:
                    V[r, c] = 0.0
                    continue
                old = V[r, c]
                # uses updated V immediately (in-place)
                V[r, c] = bellman_backup(V, (r, c), gamma=gamma)
                delta = max(delta, abs(V[r, c] - old))

        if delta < theta:
            break

    elapsed = time.perf_counter() - t0
    pi = extract_greedy_policy(V, gamma=gamma)
    return V, pi, sweeps, elapsed

def print_grid(title, grid, fmt="{:6.2f}"):
    print(title)
    for r in range(ROWS):
        row = []
        for c in range(COLS):
            row.append(fmt.format(grid[r, c]))
        print(" ".join(row))
    print()

def print_policy(title, pi):
    print(title)
    for r in range(ROWS):
        print(" ".join(f"{pi[r,c]:>4s}" for c in range(COLS)))
    print()

if __name__ == "__main__":
    print("=== Problem 3: 5×5 Gridworld Value Iteration ===\n")
    print("Reward rule (given): R(s)=+10 at goal, -5 at grey, -1 otherwise\n")
    print("Grey states:", sorted(GREY_STATES))
    print("Goal state:", GOAL)
    print("gamma =", GAMMA, "theta =", THETA)
    print()

    # Task 1: Standard Value Iteration
    V_std, pi_std, it_std, t_std = value_iteration_copy()
    print("---- Task 1: Standard (copy-based) Value Iteration ----")
    print(f"Iterations (sweeps): {it_std}")
    print(f"Time (seconds): {t_std:.6f}")
    print_grid("V* (Standard):", V_std, fmt="{:7.2f}")
    print_policy("pi* (Standard)  [ties shown as R|D etc. | G=goal]:", pi_std)

    # Task 2: In-place Value Iteration
    V_in, pi_in, it_in, t_in = value_iteration_inplace()
    print("---- Task 2: In-Place Value Iteration ----")
    print(f"Iterations (sweeps): {it_in}")
    print(f"Time (seconds): {t_in:.6f}")
    print_grid("V* (In-place):", V_in, fmt="{:7.2f}")
    print_policy("pi* (In-place):", pi_in)

    # Confirm same optimum
    same_V = np.allclose(V_std, V_in, atol=1e-6)
    same_pi = np.array_equal(pi_std, pi_in)
    print("---- Confirmation ----")
    print("Same V* ?", same_V)
    print("Same pi* ?", same_pi)

    # Simple complexity notes
    print("\n---- Complexity (simple) ----")
    print("Per sweep: O(|S||A|) where |S|=25 and |A|=4")
    print("Total: O(K|S||A|) where K = number of sweeps to converge")
    print("Memory: Standard uses 2 arrays (V and V_new); In-place uses 1 array (V)")


=== Problem 3: 5×5 Gridworld Value Iteration ===

Reward rule (given): R(s)=+10 at goal, -5 at grey, -1 otherwise

Grey states: [(0, 4), (2, 2), (3, 0)]
Goal state: (4, 4)
gamma = 0.9 theta = 0.05

---- Task 1: Standard (copy-based) Value Iteration ----
Iterations (sweeps): 9
Time (seconds): 0.000603
V* (Standard):
  -0.43    0.63    1.81    3.12    4.58
   0.63    1.81    3.12    4.58    6.20
   1.81    3.12    4.58    6.20    8.00
   3.12    4.58    6.20    8.00   10.00
   4.58    6.20    8.00   10.00    0.00

pi* (Standard)  [ties shown as R|D etc. | G=goal]:
 R|D  R|D  R|D    D    D
 R|D  R|D    R  R|D    D
   R    D  R|D  R|D    D
 R|D  R|D  R|D  R|D    D
   R    R    R    R    G

---- Task 2: In-Place Value Iteration ----
Iterations (sweeps): 9
Time (seconds): 0.000538
V* (In-place):
  -0.43    0.63    1.81    3.12    4.58
   0.63    1.81    3.12    4.58    6.20
   1.81    3.12    4.58    6.20    8.00
   3.12    4.58    6.20    8.00   10.00
   4.58    6.20    8.00   10.00    0.00

## 5×5 Gridworld (Value Iteration Results)

## 3. Results

### 3.1 Convergence

| Method | Iterations (Sweeps) | Time (seconds) |
|--------|----------------------|----------------|
| Standard Value Iteration | 9 | 0.000539 |
| In-place Value Iteration | 9 | 0.000476 |

**Observation:** In-place value iteration usually runs slightly faster because it uses a single array and immediately applies updated values. Both methods converge to the same optimal value function and policy.

---

### 3.2 Optimal Value Function \(V^*\)

| Row | Values |
|-----|--------|
| 0 | [-0.43, 0.63, 1.81, 3.12, 4.58] |
| 1 | [0.63, 1.81, 3.12, 4.58, 6.20] |
| 2 | [1.81, 3.12, 4.58, 6.20, 8.00] |
| 3 | [3.12, 4.58, 6.20, 8.00, 10.00] |
| 4 | [4.58, 6.20, 8.00, 10.00, 0.00] |

---

### 3.3 Optimal Greedy Policy \(\pi^*\)

(Notation: `R` = Right, `D` = Down, `L` = Left, `U` = Up, `G` = Goal, ties shown as `R|D`)

| Row | Policy |
|-----|--------|
| 0 | [R\|D, R\|D, R\|D, D, D] |
| 1 | [R\|D, R\|D, R, R\|D, D] |
| 2 | [R, D, R\|D, R\|D, D] |
| 3 | [R\|D, R\|D, R\|D, R\|D, D] |
| 4 | [R, R, R, R, G] |

---

## 4. In-place vs Standard Value Iteration

| Aspect | Standard | In-place |
|--------|----------|----------|
| Iterations | 9 | 9 |
| Time | 0.000539 s | 0.000476 s |
| Memory | 2 arrays (V and V_new) | 1 array (V only) |
| Convergence | Slightly slower due to copying | Slightly faster due to immediate updates |
| Complexity | \(O(|S|\cdot|A|)\) per sweep | \(O(|S|\cdot|A|)\) per sweep |

### Notes

- Both methods produce the **same optimal value function \(V^*\)** and **same optimal policy \(\pi^*\)**.
- In-place value iteration is faster in practice because it avoids copying the full value matrix.
- For larger state spaces, in-place updates reduce memory overhead and may improve convergence speed.



## Reflection
From the value table, we can clearly see that the states closer to the goal (4,4) have higher values because the agent can reach the +10 reward faster. The states far away have lower or even negative values since the agent keeps getting -1 reward while moving. The policy also shows that the best direction is mostly Right and Down, which makes sense because that is the shortest path to reach the goal.

# Problem 4

In this problem, we use Off-policy Monte Carlo with Importance Sampling to estimate the value function for the same 5×5 gridworld. Instead of calculating values directly like value iteration, Monte Carlo learns by generating many episodes using a random behavior policy. Then, importance sampling is used to correct the difference between the random behavior policy and the greedy target policy. Finally, we compare the estimated values from Monte Carlo with the optimal values obtained from value iteration.

In [7]:
# Off-policy Monte Carlo Control with Importance Sampling (Weighted IS)
# Behavior policy b: uniform random
# Target policy pi: greedy w.r.t Q

import numpy as np
import time

# -----------------------------
# 1) Environment (same as Problem 3)
# -----------------------------
ROWS, COLS = 5, 5
ACTIONS = ["R", "D", "L", "U"]
ACTION_TO_DELTA = {"R": (0, 1), "D": (1, 0), "L": (0, -1), "U": (-1, 0)}

GAMMA = 0.9
GOAL = (4, 4)
GREY_STATES = {(2, 2), (3, 0), (0, 4)}  # <-- use exactly Problem 3 grey states

def in_bounds(r, c):
    return 0 <= r < ROWS and 0 <= c < COLS

def step(state, action):
    """Deterministic move; wall => stay."""
    r, c = state
    dr, dc = ACTION_TO_DELTA[action]
    nr, nc = r + dr, c + dc
    if in_bounds(nr, nc):
        return (nr, nc)
    return (r, c)

# Reward map R(s) (state-based), but we will use reward on entering next state: r = R(s')
REWARDS = np.full((ROWS, COLS), -1.0)
for (r, c) in GREY_STATES:
    REWARDS[r, c] = -5.0
REWARDS[GOAL] = 10.0

# -----------------------------
# 2) Behavior policy b(a|s)
# -----------------------------
def behavior_policy_action():
    """Uniform random action."""
    return np.random.choice(ACTIONS)

def b_prob(a):
    """b(a|s)=1/4 for all s,a."""
    return 1.0 / len(ACTIONS)

# -----------------------------
# 3) Episode generation
# -----------------------------
def generate_episode(start_state=(0, 0), max_steps=200):
    """
    Generates an episode using behavior policy b.
    Returns list of (s, a, r) where r is reward received after taking a in s (i.e., R(s'))
    Terminal when reaching GOAL.
    """
    episode = []
    s = start_state

    for _ in range(max_steps):
        if s == GOAL:
            break
        a = behavior_policy_action()
        s_next = step(s, a)
        r = REWARDS[s_next]  # reward on entering next state
        episode.append((s, a, r))
        s = s_next

        if s == GOAL:
            break

    return episode

# -----------------------------
# 4) Off-policy MC Control with Weighted Importance Sampling
# -----------------------------
def off_policy_mc_control(num_episodes=50000, gamma=GAMMA, start_state=(0, 0), max_steps=200):
    """
    Weighted IS off-policy MC control (Sutton & Barto style).
    Learns Q and a greedy target policy pi.
    """
    Q = np.zeros((ROWS, COLS, len(ACTIONS)), dtype=float)
    C = np.zeros((ROWS, COLS, len(ACTIONS)), dtype=float)  # cumulative weights

    # deterministic greedy target policy represented as best action index per state
    pi = np.zeros((ROWS, COLS), dtype=int)  # action indices

    t0 = time.perf_counter()

    for ep in range(1, num_episodes + 1):
        episode = generate_episode(start_state=start_state, max_steps=max_steps)

        G = 0.0
        W = 1.0

        # go backward through episode
        for (s, a, r) in reversed(episode):
            sr, sc = s
            a_idx = ACTIONS.index(a)

            # return update
            G = gamma * G + r

            # weighted IS update
            C[sr, sc, a_idx] += W
            Q[sr, sc, a_idx] += (W / C[sr, sc, a_idx]) * (G - Q[sr, sc, a_idx])

            # improve target policy greedily at this state
            pi[sr, sc] = int(np.argmax(Q[sr, sc, :]))

            # if behavior action != target action, stop (because pi is deterministic greedy)
            if a_idx != pi[sr, sc]:
                break

            # update importance weight: W *= pi(a|s) / b(a|s)
            # pi(a|s)=1 for greedy action, 0 otherwise; since we didn’t break, pi(a|s)=1 here.
            W *= 1.0 / b_prob(a)

            # safety: if W becomes huge, it can cause numerical issues
            if W > 1e9:
                break

    elapsed = time.perf_counter() - t0

    # estimated state value under greedy policy: V(s) = max_a Q(s,a)
    V_est = np.max(Q, axis=2)

    # terminal value convention (fixed 0 in Problem 3/our VI script)
    V_est[GOAL] = 0.0

    return V_est, Q, pi, num_episodes, elapsed

# -----------------------------
# 5) Value Iteration baseline (for comparison)
# -----------------------------
def value_iteration(theta=0.05, gamma=GAMMA):
    V = np.zeros((ROWS, COLS), dtype=float)
    V[GOAL] = 0.0
    sweeps = 0

    t0 = time.perf_counter()
    while True:
        sweeps += 1
        V_new = V.copy()
        delta = 0.0

        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) == GOAL:
                    V_new[r, c] = 0.0
                    continue
                q_vals = []
                for a in ACTIONS:
                    nr, nc = step((r, c), a)
                    q_vals.append(REWARDS[nr, nc] + gamma * V[nr, nc])
                V_new[r, c] = max(q_vals)
                delta = max(delta, abs(V_new[r, c] - V[r, c]))

        V = V_new
        if delta < theta:
            break

    elapsed = time.perf_counter() - t0
    return V, sweeps, elapsed

# -----------------------------
# 6) Printing helpers
# -----------------------------
def print_value_table(title, V):
    print(title)
    for r in range(ROWS):
        print(" ".join(f"{V[r,c]:7.2f}" for c in range(COLS)))
    print()

if __name__ == "__main__":
    np.random.seed(0)  # reproducible

    # ----- Monte Carlo Off-policy IS -----
    N = 50000
    V_mc, Q, pi, episodes, t_mc = off_policy_mc_control(num_episodes=N)

    print("---- Off-policy Monte Carlo (Weighted IS) ----")
    print("Episodes:", episodes)
    print(f"Time (seconds): {t_mc:.6f}")
    print_value_table("Estimated V (MC):", V_mc)

    # ----- Value Iteration baseline -----
    V_vi, sweeps, t_vi = value_iteration(theta=0.05)

    print("---- Value Iteration (baseline) ----")
    print("Sweeps:", sweeps)
    print(f"Time (seconds): {t_vi:.6f}")
    print_value_table("Optimal V* (VI):", V_vi)

    # ----- Comparison summary -----
    diff = np.max(np.abs(V_mc - V_vi))
    print("---- Comparison ----")
    print(f"Max |V_MC - V*| = {diff:.4f}")

    print("\n---- Complexity notes (simple) ----")
    print("Value Iteration per sweep: O(|S||A|) with |S|=25, |A|=4")
    print("MC per episode: O(T) where T = episode length, so total O(N*T)")
    print("MC is sample-based (needs many episodes); VI is model-based (uses full transitions).")


---- Off-policy Monte Carlo (Weighted IS) ----
Episodes: 50000
Time (seconds): 27.413939
Estimated V (MC):
  -0.44    0.62    1.80    3.10    4.54
   0.62    1.80    3.11    4.56    6.17
   1.79    3.08    4.55    6.18    7.98
   3.10    4.55    6.18    7.99   10.00
   4.55    6.18    7.99   10.00    0.00

---- Value Iteration (baseline) ----
Sweeps: 9
Time (seconds): 0.000476
Optimal V* (VI):
  -0.43    0.63    1.81    3.12    4.58
   0.63    1.81    3.12    4.58    6.20
   1.81    3.12    4.58    6.20    8.00
   3.12    4.58    6.20    8.00   10.00
   4.58    6.20    8.00   10.00    0.00

---- Comparison ----
Max |V_MC - V*| = 0.0401

---- Complexity notes (simple) ----
Value Iteration per sweep: O(|S||A|) with |S|=25, |A|=4
MC per episode: O(T) where T = episode length, so total O(N*T)
MC is sample-based (needs many episodes); VI is model-based (uses full transitions).


Off-policy Monte Carlo with Importance Sampling (Results)

## 1. Setup (same as Problem 3)

- Gridworld: **5×5**
- Actions: **R, D, L, U**
- Rewards:
  - Goal state **(4,4)**: **+10**
  - Grey states **(2,2), (3,0), (0,4)**: **−5**
  - All other states: **−1**
- Discount factor: **γ = 0.9**
- Behavior policy **b(a|s)**: **uniform random** (each action probability = 1/4)
- Target policy **π(a|s)**: **greedy** (best action from learned Q)

---

## 2. Convergence / Performance

| Method | Episodes / Sweeps | Time (seconds) |
|--------|--------------------|----------------|
| Off-policy Monte Carlo (Weighted IS) | 50,000 episodes | 27.413939 |
| Value Iteration (baseline) | 9 sweeps | 0.000476 |

**Observation:** Monte Carlo took much longer because it learns from sampled episodes, while Value Iteration uses the full model and converges in a few sweeps.

---

## 3. Estimated Value Function (Monte Carlo)

Estimated \(V\) from Off-policy Monte Carlo (Weighted Importance Sampling):

| Row | Values |
|-----|--------|
| 0 | [-0.44, 0.62, 1.80, 3.10, 4.54] |
| 1 | [0.62, 1.80, 3.11, 4.56, 6.17] |
| 2 | [1.79, 3.08, 4.55, 6.18, 7.98] |
| 3 | [3.10, 4.55, 6.18, 7.99, 10.00] |
| 4 | [4.55, 6.18, 7.99, 10.00, 0.00] |

---

## 4. Optimal Value Function (Value Iteration)

Optimal \(V^*\) from Value Iteration:

| Row | Values |
|-----|--------|
| 0 | [-0.43, 0.63, 1.81, 3.12, 4.58] |
| 1 | [0.63, 1.81, 3.12, 4.58, 6.20] |
| 2 | [1.81, 3.12, 4.58, 6.20, 8.00] |
| 3 | [3.12, 4.58, 6.20, 8.00, 10.00] |
| 4 | [4.58, 6.20, 8.00, 10.00, 0.00] |

---

## 5. Accuracy Comparison

\[
\max_s |V_{MC}(s) - V^*(s)| = 0.0401
\]

This shows Monte Carlo produced a value function very close to the optimal values, but it needed many episodes and much more time.

---

## 6. Complexity (Simple)

Let \(|S|=25\) and \(|A|=4\).

### Value Iteration
- Per sweep: \(O(|S||A|)\)
- Total: \(O(K|S||A|)\), where \(K\) is sweeps until convergence

### Off-policy Monte Carlo (IS)
- Per episode: \(O(T)\), where \(T\) is episode length
- Total: \(O(NT)\), where \(N\) is number of episodes

**Key idea:**  
- Value Iteration is **model-based** (uses transitions and rewards directly).  
- Monte Carlo is **sample-based** (needs many episodes), so it is usually slower in small environments.


## Reflection

In Monte Carlo, the value function is estimated by running many random episodes and learning from experience instead of directly calculating like value iteration. The results are very close to the optimal value iteration values, which shows the algorithm is learning correctly. But Monte Carlo took a lot more time because it needs many episodes to reduce randomness and get a stable estimate.

“Monte Carlo is slower because it depends on random sampling and requires many episodes to reduce variance.”